### PR TITLE
GH-80789: Get rid of the ``ensurepip`` infra for many wheels

### PR DESCRIPTION
**Pull request by webknjaz**
This is a refactoring change that aims to simplify ``ensurepip``. Before it, this module had legacy infrastructure that made an assumption that ``ensurepip`` would be provisioning more then just a single wheel. That assumption is no longer true since [[1]][[2]][[3]].

In this change, the improvement is done around removing unnecessary loops and supporting structures to change the assumptions to expect only the bundled or replacement ``pip`` wheel.

This is a spin-off of https://github.com/python/cpython/pull/12791.

[1]: https://github.com/python/cpython/commit/ece20db
[2]: https://github.com/python/cpython/pull/101039
[3]: https://github.com/python/cpython/issues/95299

* Issue: gh-80789

### CODE REVIEW

**Comment by webknjaz** [comment]
cc @AA-Turner @pfmoore @pradyunsg

File: Lib/ensurepip/__init__.py
Comment (vstinner) [inline Lib/ensurepip/__init__.py]: If it returns a single package, it should be called _get_package(), singular, no?

File: Misc/NEWS.d/next/Library/2023-09-10-23-35-20.gh-issue-109245.yCVQbo.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2023-09-10-23-35-20.gh-issue-109245.yCVQbo.rst]: I don't think that a Changelog entry is needed. For end users, there is no visible change.

Comment (webknjaz) [reply Misc/NEWS.d/next/Library/2023-09-10-23-35-20.gh-issue-109245.yCVQbo.rst]: Don't the downstream packagers deserve it?

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: You're right. I was so focused on making the change as small as possible that I missed it. Thanks!

Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: Renamed to `_get_replacement_pip_package()`. Marking this thread as resolved.

File: Misc/NEWS.d/next/Library/2023-09-10-23-35-20.gh-issue-109245.yCVQbo.rst
Comment (webknjaz) [reply Misc/NEWS.d/next/Library/2023-09-10-23-35-20.gh-issue-109245.yCVQbo.rst]: @vstinner would you apply the corresponding labels, then?

**Comment by AA-Turner** [comment]
I've taken the liberty of pushing two changes to your branch; feel free to revert in whole or part.

Both look to simplify what ensurepip is doing internally -- e.g. with only ``pip``, having a ``_Package`` type doesn't really make sense any-more -- similarly with caching the output, which is fast to compute.

I've also removed type annotations, as these are ignored and we shouldn't introduce more (tomllib and importlib.resources are special cases).

I think there are further simplifications we could make, but I thought this was a reasonable middle-ground -- let me know your thoughts!

A

**Review by vstinner — APPROVED** [review]
LGTM. But I would prefer to get a second review on this one.

File: Lib/ensurepip/__init__.py
Comment (pfmoore) [inline Lib/ensurepip/__init__.py]: Why reverse? I guess it gives better odds of getting the correct version, but as the comment above says, we're not doing version comparison because it should never be needed.

**Review by pfmoore — APPROVED** [review]
One minor comment, but it doesn't block this - feel free to leave it if you think your version is better.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: @AA-Turner ^

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [inline Lib/ensurepip/__init__.py:36]: @AA-Turner so the reason I was using a `LookupError` was that it's better structurally and using sentinels like `None` introduces a mental burden on the callers forcing them not to forget to check for a magic value of a different type with a special meaning. This reminds me of ancient languages or golang where this same burden looks like having to check some integer return value for whether it's `-1` to decide if it's usable.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [inline Lib/ensurepip/__init__.py:25]: @AA-Turner I don't understand why you named it like this — it doesn't return a directory as far as I see. It looks like it returns a replacement package...

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [inline Lib/ensurepip/__init__.py]: @AA-Turner if we do want to include this much refactoring in the PR, why no just use my pathlib approach that reads better without looping and going into the lower-level abstractions, being shorter and easier to grasp:
```python
dist_matching_wheels = Path(_WHEEL_PKG_DIR).glob('pip-*.whl')
try:
    last_matching_dist_wheel = sorted(dist_matching_wheels)[-1]
except IndexError as index_err:
    raise LookupError from index_err
filename = last_matching_dist_wheel.name
version = filename.removeprefix('pip-').partition('-')[0]
```
_(https://github.com/python/cpython/pull/12791/files#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R197-R205)_

**Comment by webknjaz** [comment]
FreeBSD failing check reason:
> ⚠️ Failed to start an instance: FAILED_PRECONDITION: Monthly free compute limit exceeded!

**Comment by pfmoore** [comment]
> I think annotations contribute to maintainability.

I'm pretty sure there's still a policy that we don't use type annotations in the stdlib.

**Comment by webknjaz** [comment]
> I'm pretty sure there's still a policy that we don't use type annotations in the stdlib.

Interesting.. https://devguide.python.org/search/?q=type+annotations&check_keywords=yes&area=default doesn't return anything related (nor does using Google search against that site).

**Comment by AA-Turner** [comment]
> doesn't return anything related

See https://discuss.python.org/t/21487; https://github.com/python/cpython/pull/108125#discussion_r1313654070. Quoting Brett: "It's codified based on discussions among the core developers that we have all agreed to over the years"

A

File: Lib/ensurepip/__init__.py
Comment (AA-Turner) [reply Lib/ensurepip/__init__.py:25]: The function locates pip in the wheel package directory. I don't see it as "replacement" as the point of the ``--with-wheel-pkg-dir`` option is that distributors remove the bundled wheels in ``_bundled``, so there is no 'original' wheel to replace from that distributor's perspective.

File: Lib/ensurepip/__init__.py
Comment (AA-Turner) [reply Lib/ensurepip/__init__.py:36]: No strong opinions here; I see both patterns. The caller must do something with either pattern -- either check for ``None`` or catch the error. No objections to go back to the error pattern if you'd prefer.

File: Lib/ensurepip/__init__.py
Comment (AA-Turner) [reply Lib/ensurepip/__init__.py]: I noted this in the commit message but not a PR comment; sorry. The implementation of this function now short-circuits and returns the first match, but that means we need to reverse the order as previously the final match would be returned, and we want to preserve that behaviour. We retain no guarantees about finding the correct version.

File: Lib/ensurepip/__init__.py
Comment (AA-Turner) [reply Lib/ensurepip/__init__.py]: I like the ``glob`` approach (we could use ``glob.iglob`` or ``pathlib``).

File: Lib/ensurepip/__init__.py
Comment (AA-Turner) [inline Lib/ensurepip/__init__.py]: If you change to using ``pathlib``, we could leverage the traversable API a little here:

```suggestion
        if package["bundled"]:
            # Use bundled wheel package
            wheel_path = resources.files("ensurepip") / "_bundled" / wheel_name
        else:
            # Use the wheel package directory
            wheel_path  = Path(_WHEEL_PKG_DIR, wheel_name)

        whl = wheel_path.read_bytes()
        filename = Path(tmpdir, wheel_name)
        filename .write_bytes(whl)
```

File: Lib/ensurepip/__init__.py
Comment (AA-Turner) [inline Lib/ensurepip/__init__.py]: (Following on from the above if ``pathlib`` is adopted)
```suggestion
        return _run_pip([*args, "pip"], [str(filename)])
```

Comment (pradyunsg) [reply Lib/ensurepip/__init__.py]: Should that not be `os.fsdecode(filename)` instead?

Comment (AA-Turner) [reply Lib/ensurepip/__init__.py]: Perhaps? I considered using ``os.fspath``, but ``Path`` just calls ``str`` for ``__fspath__`` internally so I took the shortcut. I'm not sure on ``fsdecode`` vs ``fspath`` though.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: I think that's what I've been doing in the original PR, just with a `nullcontext` for clever code linearization...
https://github.com/python/cpython/commit/c31bdca823eab7e4ffcd70c7bd9a9895506e1e33#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R238

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: Why decode and not encode, then? Don't we pass this over the OS API boundary into another process?

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py:25]: Well, until the downloader PR is merged, there's the original "source" wheel right in the Git repo, and when distros want to replace it, they add an additional source wheel, making it two (unless they also delete the first one, of course). Hence, replacement...

**Comment by webknjaz** [comment]
@AA-Turner so I tried something which resulted in some simplification. I used pathlib but also made functions to return paths instead of complex structures with data that is never used together by the callers. I made sure that the coverage of the changed parts is at 100%, while making it less branchy...

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: This is implemented in https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b, so I'm marking the thread as resolved.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py:25]: I don't have better suggestions so I'll just mark this as resolved.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py:36]: I like forcing the caller to react rather than hoping that side won't forget... Anyway, I included this change as a part of the pathlib refactoring @ https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b. Marking the thread as resolved.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b implemented this in a more linear way. I'm marking the thread as resolved.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: Implemented as https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b. Marking as resolved.

File: Lib/ensurepip/__init__.py
Comment (pradyunsg) [inline Lib/ensurepip/__init__.py]: Use `fsdecode` for this?

File: Lib/ensurepip/__init__.py
Comment (pradyunsg) [inline Lib/ensurepip/__init__.py]: This is fairly difficult to parse and understand correctly. There's no real strong reason to make this a single expression.

Here's a suggestion I have:

```suggestion
if (_pkg_dir := sysconfig.get_config_var('WHEEL_PKG_DIR')) is not None:
    _WHEEL_PKG_DIR = Path(_pkg_dir).resolve()
else:
    _WHEEL_PKG_DIR = None
```

File: Lib/ensurepip/__init__.py
Comment (pradyunsg) [inline Lib/ensurepip/__init__.py]: I'd much rather we continue returning `None` here rather than add complexity here with the suppress.

**Review by pradyunsg — CHANGES_REQUESTED** [review]

**Comment by pradyunsg** [comment]
For the future: Please don't make additional changes that are refactoring-adjacent changes, after the PR has been reviewed and approved by multiple people.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: @pradyunsg
Could you explain why? It's made out of a pathlib object that already returns a normal unicode string. This is then injected into a snippet of Python code executed in a subprocess.

Is there a case when it'd matter? We don't work with the OS APIs on this level so I figure it'd do nothing..

Comment (pfmoore) [reply Lib/ensurepip/__init__.py]: IMO, simply because it saves people having to reason about the code. `tmp_wheel_path` is a path object, and `os.fsdecode` converts path objects to strings safely, so we know this is safe, *without* having to think too hard about edge cases like undecodable bytes, surrogate escape characters, etc.

In this case yes, both will do the same thing[^1], but why make the reader work harder than necessary?

[^1]: I assume - I haven't thought it through, which is sort of my point 🙂

Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: I'm questioning this exactly because I'm not sure it'd do the right or equivalent thing. I don't mind integrating the suggestion, just want to understand it first.

Comment (pfmoore) [reply Lib/ensurepip/__init__.py]: `os.fsdecode` will convert a bytes value correctly, whereas `str()` won't. It's not going to matter here, but `fsdecode` is, in some sense, "more correct".

In reality, though, the way to think of it is "`fsdecode` makes a filename (in whatever form you get it) into a string". And `fsencode` is the same, but to return bytes.

Honestly, though, I'm not that fussed. For this situation, `str` seems good enough even if `fsdecode` is technically better.

Comment (pradyunsg) [reply Lib/ensurepip/__init__.py]: (I've had a travel day today, and am only just seeing this)

Thanks @pfmoore for the replies here -- I'm basically on the same page. `fsdecode` more clearly communicates what the intended behaviour is here and it's _technically_ better.

> I'm not sure it'd do the right or equivalent thing

It will be equivalent for str-based `Path`s. That function takes a path / path-like object and converts it to a string representation matching the filesystem encoding (which is relevant in the case of bytes paths, but we don't have that here).

It'll return a `str` unchanged.

Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: Makes sense. Will do.

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py]: I have a different preference, but will change.

**Comment by webknjaz** [comment]
I have made the requested changes; please review again

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [inline Lib/ensurepip/__init__.py:22]: @pradyunsg converting this introduced a new uncovered line here. In https://github.com/python/cpython/pull/12791/files#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R39 I solved it by moving the config lookup from import time to runtime. What would you prefer with this?

**Comment by webknjaz** [comment]
@pradyunsg there was a merge conflict that I solved with rebase. No other changes made.

File: Lib/ensurepip/__init__.py
Comment (pradyunsg) [inline Lib/ensurepip/__init__.py:94]: This can return the constant that is used to locate the pip wheel.

```suggestion
    return _PIP_VERSION
```

File: Lib/ensurepip/__init__.py
Comment (pradyunsg) [inline Lib/ensurepip/__init__.py:53]: With that change, this function can be removed.

**Review by pradyunsg — CHANGES_REQUESTED** [review]
One last change, but LGTM otherwise!

File: Lib/ensurepip/__init__.py
Comment (pradyunsg) [reply Lib/ensurepip/__init__.py:53]: Ah, we want this to return the `WHEEL_PKG_DIR` version if used.

**Review by pradyunsg — APPROVED** [review]

File: Lib/ensurepip/__init__.py
Comment (webknjaz) [reply Lib/ensurepip/__init__.py:53]: Yep, I was making sure that the behavior is kept the same..

**Comment by webknjaz** [comment]
@pradyunsg 🛎 the CI remains green after your branch sync FYI.

**Comment by webknjaz** [comment]
@AA-Turner @pradyunsg @vstinner apply the backport labels?

**Comment by vstinner** [comment]
I don't think that such refactoring change should be backported to stable branches. Only bugfixes.

**Pull request merged**
